server/src/main/scala/org/http4s/server/middleware/RequestId.scala

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2013-2020 http4s.org
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.http4s
+package server
+package middleware
+
+import org.http4s.{Header, Http, Request, Response}
+import cats.{FlatMap, ~>}
+import cats.arrow.FunctionK
+import cats.data.{Kleisli, OptionT}
+import cats.effect.Sync
+import cats.implicits._
+import org.typelevel.ci.CIString
+import java.util.UUID
+
+/** Propagate a `X-Request-Id` header to the response, generate a UUID
+  * when the `X-Request-Id` header is unset.
+  * https://devcenter.heroku.com/articles/http-request-id
+  */
+object RequestId {
+
+  private val requestIdHeader = CIString("X-Request-ID")
+
+  def apply[G[_], F[_]](
+      fk: F ~> G,
+      headerName: CIString = requestIdHeader,
+      genReqId: Option[F[UUID]] = None
+  )(http: Http[G, F])(implicit G: FlatMap[G], F: Sync[F]): Http[G, F] = {
+    val gen = genReqId.getOrElse(F.delay(UUID.randomUUID()))
+    Kleisli[G, Request[F], Response[F]] { req =>
+      for {
+        header <- fk(req.headers.get(headerName) match {
+          case None => gen.map(reqId => Header.Raw(headerName, reqId.show))
+          case Some(header) => F.pure[Header](header)
+        })
+        response <- http(req.putHeaders(header))
+      } yield response.putHeaders(header)
+    }
+  }
+
+  def httpApp[F[_]: Sync](httpApp: HttpApp[F]): HttpApp[F] =
```

Comment:
@voidocontext was moving toward having an `httpApp` and `httpRoutes` object, and then providing overloads off of those.  Maybe that's a good idea here, particularly if we do the second constructor.

Reply:
Don't suppose you have an example? Tried doing some snooping on `@ voidocontext`, but couldn't find anything.

Made the change https://github.com/http4s/http4s/pull/3487/commits/611c898f5b7024ce0ad1c881d6f893a7b2bc4970 , this "feels" more ergonomic, but might be difficult to maintain in future.

Really appreciate the help so far! However I might need a bit more instruction (maybe method signatures) to get this PR over the line.

Reply:
He discussed it on #2402.  I don't think there are any examples yet, but I understood him to mean something like:

```scala
object httpApp {
   def apply(headerName: CIString)(httpApp: HttpApp[F]): HttpApp[F] = ???
   def apply[F[_]: Sync](
      headerName: CIString = requestIdHeader,
      genReqId: F[UUID]
  )(httpApp: HttpApp[F]): HttpApp[F] = ???
}
```

Since you're overloading `apply`, it doesn't really save much.  But if you were giving everything its own name, that technique would give you a nice namespace for the parallel methods to all be called the same thing.

Reply:
Happy to apply the same pattern that we're going to use for all the other middleware.

Reply:
Apologies for the delay, I got swamped with other things, but indeed this is what I had in mind when I suggested this approach.